### ISSUE TITLE
Move cache and log directories out of application module

### ISSUE THREAD

**Issue by lenton** [kohana/kohana#62]
I propose both the cache and log directories get moved out of the application module and placed into a `var` directory (or similar).

The cache and log files currently reside in the cascading filesystem which is not where they belong. These files are not intended to be overridden by modules with higher precedence.

A benefit to having log and cache contained in the same directory is that they can be found and their permissions changed trivially.

**Comment by enov** [comment]
:+1:

**Comment by acoulton** [comment]
:+1:

**Comment by rjd22** [comment]
I think this is not a good change. There can be files in the cache directory that are used by the application. Also if we take look at symfony the cache and log directory are also in the app folder. 

Please lets not make our choices by basing them on the CFS since we might not want to replace it for namespaces. 

What problem are you trying to solve?

**Comment by lenton** [comment]
> There can be files in the cache directory that are used by the application.

There shouldn't be anything other than generated cache files in the cache directory; if you can't safely delete the contents of it then you're doing it wrong.

> Please lets not make our choices by basing them on the CFS since we might not want to replace it for namespaces.

This is a change that can be made now, it doesn't rely on my kohana/modules package.

> What problem are you trying to solve?

It's an enhancement of the folder structure, the benefits are:
- 'variable directories' are clearly marked and secluded from application files.
- Setting up your project to use the server's `/var/cache` and `/var/log` is straight forward.
- You can freely use 'cache' and 'log' directories as normal in the CFS if you want.
- It's one step closer towards the application module being optional.

**Comment by rjd22** [comment]
I'm still not convinced this is a change we should make. It doesn't solve any big issues and I think trying to remove the application module might be wrong. I would like to see @shadowhand his opinion on this.

**Comment by acoulton** [comment]
Removing the application module is a separate discussion (and I'm not convinced on that either).

But I agree 'cache' and 'log' aren't part of the application's resources (classes, views, config, etc) and they uniquely need writeable storage - putting them in a 'var' gives a more obvious hint that any app-specific writeable directories (uploads, etc) should probably also be in there. It's also easier to manage deployments, gitignores, etc if everything writeable/persistent is all in a separate place to the deployed code.

That said, either way this isn't a huge issue - both paths are configured in the bootstrap, and this project just provides a sample - nobody actually uses it in real apps. It's easy for end-users to ignore this change, or for them to use eg '/var/cache' instead.

But I think better for our sample to promote a cleaner setup if possible.

**Comment by shadowhand** [comment]
I agree with @rjd22, there is no valid reason for this change, it appears to be "change for the sake of change" in my opinion. You can already change the paths via configuration, but having them in `application/` is a sane default.

**Comment by enov** [comment]
@rjd22

> if we take look at symfony the cache and log directory are also in the app folder

Symfony3 defines `var` folder outside the app folder similar to the one proposed here, among other folder changes. You can also use the Symfony2 app with Symfony3 folder specs as the spec is BC and folders configurable. So in the long run we are getting closer to symfony and not the other way around.

@shadowhand 

it's more about collecting those writable dirs into a single folder, rather than whether it is outside or inside of the `application` dir. It is about to give the permissions/rights once to the parent folder. Also it would be easier to gitignore. And as a developer you might want to add additional folders like your application `data` or `uploads` folder more easily in there, by just creating the directories manually or at runtime (which is now possible because the parent folder is writable).

**Comment by acoulton** [comment]
I don't think having separate writeable directories under `application` is a sane default, for reasons already outlined.

Agree we want to avoid lots of small unimportant changes, but the threshold is lower on this repo because it's just showing recommended practice, there is no BC concern.

If this was a PR I'd have merged it by now, it's not worth discussion IMHO.
